### PR TITLE
[enhancement](Nereids) add all necessary PhysicalDistribute on Join's child and update cost model

### PR DESCRIPTION
**Pull request by morrySnow**
# Proposed changes

In an earlier PR #11976 , we add shuffle join and bucket shuffle support. But if join's right child's distribution spec satisfied join's require, we do not add distribute on right child. Instead of, do it in plan translator.
It is hard to calculate accurate cost in this way, since we some distribute cost do not calculated.
In this PR, we introduce a new shuffle type BUCKET, and change the way of add enforce to ensure all necessary distribute will be added in cost and enforcer job.

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by jackwener — DISMISSED** [review]
LGTM general.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/cascades/CostAndEnforcerJob.java [hunk 1]
```diff
@@ -184,33 +186,47 @@ public void execute() {
...
-                enforce(outputProperty, requestChildrenProperty);
+                // colocate join and bucket shuffle join cannot add enforce to satisfy required property.
+                if (enforce(outputProperty, requestChildrenProperty)) {
 
-                if (curTotalCost < context.getCostUpperBound()) {
-                    context.setCostUpperBound(curTotalCost);
+                    // record map { outputProperty -> outputProperty }, { ANY -> outputProperty },
+                    recordPropertyAndCost(groupExpression, outputProperty, outputProperty, requestChildrenProperty);
+                    recordPropertyAndCost(groupExpression, outputProperty, PhysicalProperties.ANY,
+                            requestChildrenProperty);
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/cascades/CostAndEnforcerJob.java:196]: enforce return false means this branch is not work

**Comment by morrySnow** [comment]
need to refactor whole enforce. need some time

**Review by jackwener — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java [hunk 2]
```diff
@@ -106,19 +112,47 @@ public CostEstimate visitPhysicalLocalQuickSort(
...
+            }
 
+            // replicate
+            if (spec instanceof DistributionSpecReplicated) {
+                int beNumber = ConnectContext.get().getEnv().getClusterInfo().getBackendIds(true).size();
+                int instanceNumber = ConnectContext.get().getSessionVariable().getParallelExecInstanceNum();
+                beNumber = Math.max(1, beNumber);
+
+                return new CostEstimate(
+                        childStatistics.computeSize() * beNumber,
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java:138]: do not multiply by beNumber. In fact, for replicate mode, the beNumber is bigger, the workload is smaller.
That means we need cost of a single BE node, not the cost of cluster.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java [hunk 3]
```diff
@@ -158,22 +192,10 @@ public CostEstimate visitPhysicalNestedLoopJoin(
...
-                        rightStatistics.computeColumnSize(rightIds),
-                        0);
-            }
 
-            // TODO: network 0?
             return new CostEstimate(
-                    (leftStatistics.computeColumnSize(leftIds) + rightStatistics.computeColumnSize(rightIds)) / 2,
-                    rightStatistics.computeColumnSize(rightIds),
+                    leftStatistics.computeSize() * rightStatistics.computeSize(),
+                    rightStatistics.computeSize(),
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java:198]: In my understanding, the memory is the size of two blocks, from left child and right child respectively.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java (hunk 2 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java:138]: This PR aim to get a correct plan of colocate join and bucket shuffle join. Good cost model is not including. We need to adjust it finely after this PR and #12767

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java (hunk 3 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostCalculator.java:198]: This PR aim to get a correct plan of colocate join and bucket shuffle join. Good cost model is not including. We need to adjust it finely after this PR and https://github.com/apache/doris/pull/12767

**Review by jackwener — APPROVED** [review]
LGTM

**Pull request merged**